From 438410b26dc52d9156a22dfd54becb2d2a1fd442 Mon Sep 17 00:00:00 2001
Message-Id: <438410b26dc52d9156a22dfd54becb2d2a1fd442.1278011888.git.linz@li-pro.net>
In-Reply-To: <8a9838187997476085cc0cbc1749cf5e373e2115.1278011888.git.linz@li-pro.net>
References: <8a9838187997476085cc0cbc1749cf5e373e2115.1278011888.git.linz@li-pro.net>
From: Stephan Linz <linz@li-pro.net>
Date: Thu, 1 Jul 2010 20:52:45 +0200
Subject: [PATCH 2/2] net: on microblaze declare cpu_eth_init() as external

It seams the current microblaze compiler (4.1.2) is not doing a good job
of handling the "weak" attribute of gcc. Read more at:
  http://www.mail-archive.com/u-boot@lists.denx.de/msg09779.html
  http://forums.xilinx.com/t5/Embedded-Linux/Problem-Building-Linux-with-PCI-for-Microblaze/m-p/67037

The only thing we can do is to declare cpu_eth_init() as external.

Signed-off-by: Stephan Linz <linz@li-pro.net>
---
 net/eth.c |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/net/eth.c b/net/eth.c
index 83d559c..38de2ba 100644
--- a/net/eth.c
+++ b/net/eth.c
@@ -78,7 +78,19 @@ static int __def_eth_init(bd_t *bis)
 {
 	return -1;
 }
+#if defined(__MICROBLAZE__) && \
+	(__GNUC__ <= 4) && (__GNUC_MINOR__ <= 1) && (__GNUC_PATCHLEVEL__ <= 3)
+/* It seams the current microblaze compiler (4.1.2) is not doing a good job
+ * of handling the "weak" attribute of gcc. Read more at:
+ *   http://www.mail-archive.com/u-boot@lists.denx.de/msg09779.html
+ *   http://forums.xilinx.com/t5/Embedded-Linux/Problem-Building-Linux-with-PCI-for-Microblaze/m-p/67037
+ *
+ * The only thing we can do is to declare cpu_eth_init() as external.
+ */
+extern int cpu_eth_init(bd_t *bis);
+#else
 int cpu_eth_init(bd_t *bis) __attribute__((weak, alias("__def_eth_init")));
+#endif
 int board_eth_init(bd_t *bis) __attribute__((weak, alias("__def_eth_init")));
 
 extern int mv6436x_eth_initialize(bd_t *);
-- 
1.6.0.4

